clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -1223,6 +1233,26 @@ public void maybeAutoCommitOffsetsAsync(long now) {
         }
     }
 
+    private boolean waitForPendingAsyncCommits(Timer timer) {
```

Comment:
The code appears to check `inFlightAsyncCommits` only in the case where the incoming `offsets` map is empty. Wouldn't we want to always check that?

@philipnee's suggestion makes sense to me:

```java
if (offsets.isEmpty() && inflightCommit.get() == 0)
    return true;
```

`sendOffsetCommitRequest` already handles the case where the incoming map is empty, so perhaps that could be left as is?

Reply:
I tried to apply @philipnee's suggestion but it will be very tricky; logic will spread around and understandability will suffer a lot. Method commitSync must either return false, or wait until the pending async commits have been handled. I discovered that the test covers this aspect well. When we use the if statement and then use the rest of the code as is, it will return true (because we gave it an empty offsets) even though there are still inflight async commits. To fix this, more logic is needed. How to do this exactly is beyond me at the moment. I would prefer to not change this aspect.

Reply:
could you elaborate `logic will spread around and understandability will suffer a lot` : I was suggesting to first check if there's any inflight commits, then wait for the coordinator to be connected.  After the connection is established, you can try to send the async commits.

Reply:
I am struggling, so many words I do not know the meaning of. Because of that I do not understand your suggestion. What is a coordinator, what does it mean to be connected? Why do you want to send async commits from the commitSync method?

> could you elaborate logic will spread around and understandability will suffer a lot
For example, `sendOffsetCommitRequest` does nothing when it finds an empty offsets (it returns early). So that method has to be changed. I stopped there because I do not understand what this method does (definitely more than sending requests 😄). At first I though we may need to check the inflighAsyncCommits there, but perhaps we only have to remove the empty check on top of `sendOffsetCommitRequest`.

Reply:
the sendOffsetCommitRequest method basically 1. if the offset is empty, complete the future and return immediately, 2. if there's no coordinator `checkAndGetCoordinator` then complete the future with exception (like completeExceptionally) and return, otherwise 3. create the request data and send it `client.send(coordinator, builder).compose(new OffsetCommitResponseHandler(offsets, generation));` 

What I'm suggesting is not to change that method, but to change the `commitOffsetsSync`.  My suggestions are:

Currently, it returns immediately if the offset is empty. But we don't want that, because we also want to check if there's any inflightAsyncCommits. 

Now, if the we can't return immediately, we will need to send these commits, and the requirement is to `coordinatorUnknownAndUnreadySync`.  This check is already in place, so I think you just need to try to send the async commit after this check.

So there's not very much code change there :) 

The main concern me and David have is managing 2 atomic int, because at certain point one might forget to update one of it and causes some weird bug.

Reply:
> so I think you just need to try to send the async commit after this check.

I thought that the async commits would already have been send and that we only need to somehow communicate with the broker to see if they have been completed (and if so, call the callbacks). Is that assumption wrong?

Reply:
> The main concern me and David have is managing 2 atomic int, because at certain point one might forget to update one of it and causes some weird bug.

Yes, I share that concern. I do not have a good solution though.

Reply:
Hey @erikvanoosten - I took a closer look at your PR. I think your current approach is fine.

Reply:
> The code appears to check `inFlightAsyncCommits` only in the case where the incoming `offsets` map is empty. Wouldn't we want to always check that?

@kirktrue I think it was changed this way because when offsets is not empty, method `invokeCompletedOffsetCommitCallbacks();`is called which already does what we need.